feat: remove nuxt-loading div from DOM when not used

pimlie opened the pull request:
feat: split nuxt-loading.vue in js/css files
feat: use css classes instead of style props
feat: add loading.css property to fully ignore the default css

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] New feature (non-breaking change which adds functionality)

## Description
Remove the nuxt-loading div from the DOM when its not used and fully use css-classes instead of style properties. Also add a `loading.css` boolean property so you can fully ignore the default nuxt css.

## Checklist:
<!--- Put an `x` in all the boxes that apply. -->
<!--- If your change requires a documentation PR, please link it appropriately -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: https://github.com/nuxt/docs/pull/924)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests passed.

Will add a docs PR if this gets approved.

pimlie committed:
feat: remove nuxt-loading div from DOM when not used

feat: split nuxt-loading.vie in js/css files

feat: add css classes instead of style props

feat: add loader.css property

galvez reviewed (changes requested):
@pimlie take a look at the failing test -- LGTM otherwise

pimlie committed:
fix: update new extract-css hash

pimlie commented:
@galvez Yeah sorry, I wasnt sure whether I caused that issue and I didnt want to touch things I did not specifically change. But looking at it now I understand why ;)

galvez commented:
@pimlie Can you also add a test that determines it's not being included when not used?

pimlie committed:
add loading/progressbar test

pimlie committed:
lint fixes

pimlie committed:
fix test

pimlie commented:
I 'fixed' the tests on circleci by adding a `Utils.waitFor(100)` at the start of the new test. It works now, but  this could possibly trigger intermittent errors when the circleci servers are busy. Its not a perfect solution but I guess the tests are quicker triggered then puppeteer / chrome can handle all DOM events and I cant think of another solution atm to wait for the DOM events to finish.

Let me know if anyone has a better solution, maybe otherwise just put this test in a separate test file?

pi0 commented on lib/app/components/nuxt-loading.js:
```diff
@@ -1,43 +1,47 @@
 
 export default {
   name: 'nuxt-loading',
+  render(h) {
```
Render function instead of`<template>` seems a good idea for this 👍

pimlie commented on lib/app/App.js:
```diff
@@ -1,5 +1,5 @@
 import Vue from 'vue'
-<% if (loading) { %>import NuxtLoading from '<%= (typeof loading === "string" ? loading : "./components/nuxt-loading.vue") %>'<% } %>
+<% if (loading) { %>import NuxtLoading from '<%= (typeof loading === "string" ? loading : "./components/nuxt-loading.js") %>'<% } %>
```
Would you be ok with omitting the file extension in that case? I am not fully sure if that will always work, but in my experience omitting the file extensions resolves in either `.vue` or `.js`.

pimlie commented on lib/app/components/nuxt-loading.js:
```diff
@@ -1,43 +1,47 @@
-
-<script>
-import Vue from 'vue'
+<% if (loading && loading.css) { %>import "./nuxt-loading.css"<% } %>
```
Will revert this back :)

Atinux commented:
I agree with @pi0 requested changes, looking forward to this PR :)

pimlie committed:
revert back to SFC

pimlie committed:
rewrite extract css test to remove [hash] dep

pimlie committed:
Merge branch 'dev' into feat-render-loading

pimlie committed:
clean-up nuxt-loading even more

add nuxt-loading component tests with vue-server-renderer

revert flaky e2e progressbar test

add nuxt.options.loading.throttle option with defualt value 200

pimlie commented:
This PR is ready for review again. These last changes introduce one breaking change: before if you paused loading with `$loading.pause()` you could / needed to restart loading again by calling `$loading.start()`. This has now changed to `$loading.resume()`. If you call `start()` the loading will always start from 0 again. Hope you agree with me thats better as otherwise you also have the throttle timeout after resuming every pause :)

If you wish I can work on this tomorrow, but the e2e tests are probably not as they should be. E.g. the basic browser `/noloading` test actually errors out to a nuxt-error page which is not detected at all. We should add a listener for errors to fully make sure there are none.
Btw, the /noloading page errors out because `this.$nuxt.$loading` is not yet available when mounted() is called. It is only available after a nextTick(). Also think the start/finish calls are reversed. @galvez could you maybe confirm / have a look at this?

Not sure why the circleci:audit test failed, looks like a temporary CI issue?

pimlie committed:
fix throttle test

add pause/resume test

Commit:
fix: Use vue ext

Atinux commented on test/e2e/basic.browser.test.js:
```diff
@@ -28,10 +28,12 @@ describe('basic browser', () => {
-    const { hook } = await page.nuxt.navigate('/noloading')
-    const loading = await page.nuxt.loadingData()
+    const { hook } = await page.nuxt.navigate('/noloading', false)
+    await Utils.waitFor(nuxt.options.loading.throttle + 100)
```
Could we maybe move this right into `page.nuxt.navigate` @pimlie ?

pimlie replied:
Yes, or we could add a call to `page.waitForFunction` instead. But as mentioned I actually think the e2e tests need some more refactoring anyway as there doesnt seem to be enough protection against unexpected exceptions at the moment. Maybe we can change this with that refactoring?

Atinux commented:
Please fix pending conflicts so we can work further on this :)

pimlie committed:
Merge branch 'dev' into feat-render-loading

pimlie committed:
more features and fixes

feat: new config option 'continuous'
If true the progress bar keeps changing also after duration time has
passed. It switches from ltr growing to rtl shrinking, then ltr growing
again with ltr: false

fix: remove Math.random() for percent increase, it caused choppy
behaviour and more importantly when loading took exactly duration time
there was a 100% chance the progress bar was not yet at 100%

fix: add safeguards so percent can never be <0 or >100

fix: calculate _cut correctly for duration < 1000

fix: remove automatically finishing when >95% is reached. This behaviour was
counter-intuitive, when duration time had almost passed duration the
progress bar always finished, regardless whether it was truly finished or
not. Now the progress bar stays at 100% until loading really finishes
(with continuous = false)

feat: add nuxt-progess-notransition class to disable transitions when we
are switching from grow to shrink

pimlie committed:
Merge branch 'dev' into feat-render-loading

pimlie committed:
Merge branch 'feat-render-loading' of into feat-render-loading

pimlie committed:
Merge branch 'dev' into feat-render-loading

pimlie commented:
Please see the message of commit e7bddc9 for all the details.

And I am not sure what the idea behind PR 3953 was, but as git wasnt cooperating it took me 20mins to merge the changes back in I had already fixed days before in this PR. Double work, double fun I guess? :roll_eyes: :wink:

pimlie committed:
revert fix for _cut which wasnt broken at all

i just like to do things more difficult :s

pimlie committed:
Merge branch 'dev' into feat-render-loading

pimlie committed:
Merge branch 'dev' into feat-render-loading

pimlie committed:
fix: change waitFor import

fix: add yarn.lock after vue-jest install

pimlie committed:
fix: lint and import in e2e

pi0 commented on packages/common/src/nuxt.config.js:
```diff
@@ -194,7 +194,9 @@ export default {
     throttle: 200,
     duration: 5000,
-    rtl: false
+    continuous: false,
```
Docs should be updated to add `continuous` too

pimlie committed:
fix: explicitly set left style, false doesnt remove it once set

pimlie committed:
Merge branch 'dev' into feat-render-loading

pimlie committed:
fix: remove nuxt.config from common

pimlie committed:
fix: remove unnecessary decimal test

pimlie commented:
I have added a PR for the missing docs (you can also check that PR for a gif of the continuous loading)

pimlie committed:
fix: wait longer for azure

pi0 reviewed (approved):
LGTM

manniL commented on packages/app/template/components/nuxt-loading.vue:
```diff
@@ -1,115 +1,177 @@
-      color: '<%= loading.color %>',
-      failedColor: '<%= loading.failedColor %>'
+      duration: <%= loading.duration || 3000 %>,
+      continuous: <%= loading.continuous || false %>
```
Do we need the default fallback here? I think we already have them in the config. Might be worth a refactor while the PR is up :thinking:

pimlie replied:
This is a quick fix / workaround to prevent lint errors with `lint:app` in the with-config fixture. If you provide your own loading component, nuxt-loading.vue is still being generated by the builder but without any loading properties as `loading: false`.
